examples/game_of_tag/tag_adapters.py

```diff
@@ -0,0 +1,230 @@
+        # make the last observation bad for prey to discourage off road
+        states += [
+            {
+                "heading": np.array([0]),
+                "speed": np.array([0]),
+                "position": ego_state.position[:2],
+                "lane_index": ego_state.lane_index,
+                "distance": np.array([COLLIDE_DISTANCE]),  # give max reward to predator
+            }
+        ] * (len(wanted_ids) - len(states))
+    elif wanted_ids == PREDATOR_IDS:
+        # ego is prey, predator went off road
+        # make the last observation bad for predator
+        states += [
+            {
+                "heading": np.array([0]),
+                "speed": np.array([0]),
+                "position": np.array([1000, 1000]),
+                "lane_index": ego_state.lane_index,
+                "distance": np.array([1e3 - 1]),  # makes  position far from predator
+            }
+        ] * (len(wanted_ids) - len(states))
+
+    return states
+
+
+def min_distance_to_rival(ego_position, rival_ids, neighbour_states):
+    rival_vehicles = filter(
+        lambda v: _is_vehicle_wanted(v.id, rival_ids), neighbour_states
+    )
+    rival_positions = [p.position for p in rival_vehicles]
+
+    return min(
+        [np.linalg.norm(ego_position - prey_pos) for prey_pos in rival_positions],
+        default=0,
+    )
+
+
+def observation_adapter(observations):
+    nv_states = observations.neighborhood_vehicle_states
+    ego = observations.ego_vehicle_state
+
+    target_vehicles = None
+    if _is_vehicle_wanted(ego.id, PREY_IDS):
+        target_vehicles = get_specific_vehicle_states(nv_states, PREDATOR_IDS, ego)
+    elif _is_vehicle_wanted(ego.id, PREDATOR_IDS):
+        target_vehicles = get_specific_vehicle_states(nv_states, PREY_IDS, ego)
+
+    return {
+        "heading": np.array([ego.heading]),
+        "speed": np.array([ego.speed]),
+        "position": np.array(ego.position[:2]),
+        "lane_index": ego.lane_index,
+        "target_vehicles": tuple(target_vehicles),
+    }
+
+
+def dominant_reward(distance):
+    if distance == COLLIDE_DISTANCE:
+        return 10
+    return min(0.5 / ((distance - COLLIDE_DISTANCE) ** 2), 10)
+
+
+def predator_reward_adapter(observations, env_reward_signal):
+    rew = 0
+    ego = observations.ego_vehicle_state
+
+    # Primary reward
+    distance_to_target = min_distance_to_rival(
+        ego.position,
+        PREY_IDS,
+        observations.neighborhood_vehicle_states,
+    )
+
+    rew += dominant_reward(distance_to_target)
+
+    events = observations.events
+    for c in observations.events.collisions:
+        if _is_vehicle_wanted(c.collidee_id, PREY_IDS):
+            rew += global_rewards.collesion_with_target
+            print(
+                f"predator {ego.id} collided with prey {c.collidee_id} distance {distance_to_target}"
+            )
+        # # keeping this commented code for expanding to mutiple preys and predators in the future
```

Comment:
I think we can pass by with 1 prey and 1 predator for now but we would want to do a follow-up that implements multiple prey and predators. Please make an issue for that improvement.